third_party/mbedtls/CMakeLists.txt

```diff
@@ -26,73 +26,57 @@
 #  POSSIBILITY OF SUCH DAMAGE.
 #
 
-add_library(mbedcrypto)
-
-set(OT_MBEDTLS_DEFAULT_CONFIG_FILE "\"mbedtls-config.h\"")
+set(OT_MBEDTLS_DEFAULT_CONFIG_FILE \"${CMAKE_CURRENT_SOURCE_DIR}/mbedtls-config.h\")
 
 set(OT_MBEDTLS_CONFIG_FILE "" CACHE STRING "The mbedTLS config file")
 
-target_compile_definitions(mbedcrypto
+set(ENABLE_TESTING OFF CACHE BOOL "Disable mbedtls test")
+set(ENABLE_PROGRAMS OFF CACHE BOOL "Disable mbetls program")
+
+add_subdirectory(repo)
+
+## Include the user config file by absolute path to avoid exposing current directory.
+target_compile_definitions(mbedtls
     PUBLIC
         "MBEDTLS_CONFIG_FILE=$<IF:$<BOOL:${OT_MBEDTLS_CONFIG_FILE}>,${OT_MBEDTLS_CONFIG_FILE},${OT_MBEDTLS_DEFAULT_CONFIG_FILE}>"
     PRIVATE
         ${OT_PRIVATE_DEFINES}
 )
-
-target_include_directories(mbedcrypto
+target_include_directories(mbedtls
     PUBLIC
-        ${PROJECT_SOURCE_DIR}/third_party/mbedtls/repo/include
+        ${CMAKE_CURRENT_SOURCE_DIR}/repo/include
     PRIVATE
         ${OT_PUBLIC_INCLUDES}
         ${OT_PRIVATE_INCLUDES}
+        ${PROJECT_SOURCE_DIR}/src/core
```

Comment:
I think this needs to be `PUBLIC`. Because the `mbedtls` target includes `mbedtls-config.h` in its public headers (header files in `${CMAKE_CURRENT_SOURCE_DIR}/repo/include`), and `mbedtls-config.h` includes a core config file in `${PROJECT_SOURCE_DIR}/src/core`. So targets depending on `mbedtls` will not be able to find the core config file if we make `${PROJECT_SOURCE_DIR}/src/core` PRIVATE.

even worse, `${OT_PRIVATE_DEFINES}` below also needs to be PUBLIC. Otherwise, targets other than `openthread-ftd` and `openthread-mtd` will see different mbedtls header files.

To avoid exposing `${PROJECT_SOURCE_DIR}/src/core` and `${OT_PRIVATE_DEFINES}`, one option is to provide mbedtls macros by `target_compile_definitions(...)` rather than `MBEDTLS_CONFIG_FILE`. thoughts?

Reply:
This seems to work as is. Note that `MBEDTLS_CONFIG_FILE` is PUBLIC with absolute path. Thoughts?

Reply:
Yes,  `MBEDTLS_CONFIG_FILE` has absolute path, so we don't need to include `${CMAKE_CURRENT_SOURCE_DIR}` to find `mbedtls-config.h`. But `mbedtls-config.h` is not self-contained, because it includes `penthread-core-config.h`.

It works now, because we manually add the required definitions to all targets (e.g. openthread-ftd) that depends on the mbedTLS targets. But I don't think this is how modern CMake should work. Ideally, I would expect any target `A` depending on target  `B` just need to do `target_link_libraries(A PRIVATE B)` and inherit include directories and definitions, but don't need to manually add them. thoughts?

Reply:
The dependency on `openthread-core-config.h` generally applies to all targets in the OpenThread project. It's not clear to me that the medtls target should be the one responsible for exporting where `openthread-core-config.h` is located.

Reply:
Agree that mbedTLS targets should not export the location of `openthread-core-config.h`. But we are exporting `MBEDTLS_CONFIG_FILE ` and it includes `openthread-core-config.h`, which force we to export location of `openthread-core-config.h` to make it work with modern cmake.

I think currently it is ok. But make openthread works with modern cmake and easier to be used by applications, we need to fix this in the future.

Reply:
OK, if I understand correctly, the real problem is the inclusion of `openthread-core-config.h` in `mbedtls-config.h` and you'd like to see that removed. That makes sense to me. At the same time, `openthread-core-config.h` is what determines what features are needed by mbedTLS. Given that this is an existing issue not related to this PR, I'd prefer to address this separately.

Reply:
Agree.